File: JULY-2025/10-07-2025/p3440.dart
class Solution {
  int maxFreeTime(int eventTime, List<int> startTime, List<int> endTime) {
    int n = startTime.length;
    if (n == 0) return eventTime;

    List<int> gaps = List.filled(n + 1, 0);
    gaps[0] = startTime[0];

    for (int i = 1; i < n; i++) {
      gaps[i] = startTime[i] - endTime[i - 1];
    }
    gaps[n] = eventTime - endTime[n - 1];

    List<int> largestRight = List.filled(n + 1, 0);
    for (int i = n - 1; i >= 0; i--) {
      largestRight[i] = max(largestRight[i + 1], gaps[i + 1]);
    }

    int maxFreeTimeResult = 0;
    int largestLeft = 0;

    for (int i = 1; i < gaps.length; i++) {
      int meetingDuration = endTime[i - 1] - startTime[i - 1];
      bool canFitLeft = largestLeft >= meetingDuration;
      bool canFitRight = largestRight[i] >= meetingDuration;

      if (canFitLeft || canFitRight) {
        int merged = gaps[i - 1] + gaps[i] + meetingDuration;
        maxFreeTimeResult = max(maxFreeTimeResult, merged);
      }

      int current = gaps[i - 1] + gaps[i];
      maxFreeTimeResult = max(maxFreeTimeResult, current);
      largestLeft = max(largestLeft, gaps[i - 1]);
    }

    return maxFreeTimeResult;
  }
}

File: JULY-2025/10-07-2025/p3440.rs
use std::cmp::max;

impl Solution {
    pub fn max_free_time(event_time: i32, start_time: Vec<i32>, end_time: Vec<i32>) -> i32 {
        let n = start_time.len();
        if n == 0 {
            return event_time;
        }

        let mut gaps = vec![0; n + 1];
        gaps[0] = start_time[0];
        for i in 1..n {
            gaps[i] = start_time[i] - end_time[i - 1];
        }
        gaps[n] = event_time - end_time[n - 1];

        let mut largest_right = vec![0; n + 1];
        for i in (0..n).rev() {
            largest_right[i] = max(largest_right[i + 1], gaps[i + 1]);
        }

        let mut max_free = 0;
        let mut largest_left = 0;

        for i in 1..=n {
            let duration = end_time[i - 1] - start_time[i - 1];
            let can_fit_left = largest_left >= duration;
            let can_fit_right = largest_right[i] >= duration;

            if can_fit_left || can_fit_right {
                let merged = gaps[i - 1] + gaps[i] + duration;
                max_free = max(max_free, merged);
            }

            max_free = max(max_free, gaps[i - 1] + gaps[i]);
            largest_left = max(largest_left, gaps[i - 1]);
        }

        max_free
    }
}